zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/message/CorrelateMessageTest.java

```diff
@@ -0,0 +1,450 @@
+    deployAndStartProcessWithIntermediaryMessageEvent(messageName, correlationKey);
+
+    // when
+    engine
+        .messageCorrelation()
+        .withName(messageName)
+        .withCorrelationKey(correlationKey)
+        .correlate();
+
+    // then
+    assertThat(
+            RecordingExporter.records()
+                .limit(r -> r.getIntent().equals(MessageIntent.EXPIRED))
+                .filter(
+                    r ->
+                        List.of(
+                                ValueType.MESSAGE_CORRELATION,
+                                ValueType.MESSAGE_SUBSCRIPTION,
+                                ValueType.PROCESS_MESSAGE_SUBSCRIPTION,
+                                ValueType.MESSAGE)
+                            .contains(r.getValueType())))
+        .extracting(Record::getIntent)
+        .containsExactly(
+            MessageCorrelationIntent.CORRELATE,
+            MessageIntent.PUBLISHED,
+            MessageSubscriptionIntent.CORRELATED,
+            MessageIntent.EXPIRED,
+            ProcessMessageSubscriptionIntent.CORRELATE,
+            ProcessMessageSubscriptionIntent.CORRELATED,
+            MessageSubscriptionIntent.CORRELATE,
+            MessageSubscriptionIntent.CORRELATED,
+            MessageCorrelationIntent.CORRELATED);
+  }
+
+  @Test
+  public void shouldCorrelateMessageToStartEvent() {
+    // given
+    deployProcessWithMessageStartEvent();
+    final var correlationKey = "correlationKey";
+
+    // when
+    final var record =
+        engine
+            .messageCorrelation()
+            .withCorrelationKey(correlationKey)
+            .withName("messageName")
+            .correlate();
+
+    // then
+    assertMessageIsCorrelated(record);
+  }
+
+  @Test
+  @Ignore("Enable in https://github.com/camunda/camunda/issues/20175")
+  public void shouldCorrelateToMessageIntermediaryEvent() {
+    // given
+    final var messageName = "messageName";
+    final var correlationKey = "correlationKey";
+    deployAndStartProcessWithIntermediaryMessageEvent(messageName, correlationKey);
+
+    // when
+    final var record =
+        engine
+            .messageCorrelation()
+            .withName(messageName)
+            .withCorrelationKey(correlationKey)
+            .correlate();
+
+    // then
+    assertMessageIsCorrelated(record);
+  }
+
+  @Test
+  @Ignore("Enable in https://github.com/camunda/camunda/issues/20175")
+  public void shouldCorrelateToMessageBoundaryEvent() {
+    // given
+    final var messageName = "messageName";
+    final var correlationKey = "correlationKey";
+    deployAndStartProcessWithMessageBoundaryEvent(messageName, correlationKey);
+
+    // when
+    final var record =
+        engine
+            .messageCorrelation()
+            .withName(messageName)
+            .withCorrelationKey(correlationKey)
+            .correlate();
+
+    // then
+    assertMessageIsCorrelated(record);
+  }
+
+  @Test
+  public void shouldNotCorrelateMessageIfNoProcess() {
```

Comment:
❓ What's the difference between this test and `shouldNotCorrelateWhenNoSubscriptions` ?

Reply:
The `shouldNotCorrelateWhenNoSubscriptions` test only checks some commands/events have been written. This one checks some more specifics about the event, such as message name and correlation key.